File: bench/branches.lua
-- The Lua twin of branches.sl.
--
-- Lua has no `match`, so the twin is the same chain of comparisons `dispatch.lua` writes -- and no
-- `continue`, so the early exit is Lua's own idiom for it, a `goto` to a label at the bottom of the
-- loop body. `pcall` is Lua's `try`: it returns `true` plus the call's result on success and `false`
-- plus the error object on failure, and `error()` is `throw`.
--
-- Lua 5.5's integers are 64-bit, and every value here stays under 2^31 with the one product taken
-- (`seed * 16807`) under 2^53, so this is exact integer arithmetic throughout and prints the same
-- answer as the other three.

local function risky(r)
    if r == 0 then error("boom") end
    return 0
end

local function run()
    local seed = 1
    local i = 0
    local flag_count = 0
    local chain_total = 0
    local nested_count = 0
    local continue_count = 0
    local expr_sum = 0
    local match_total = 0
    local fault_count = 0

    while i < 2000000 do
        seed = (seed * 16807) % 2147483647

        local r = seed % 1000

        i = i + 1

        if r % 7 == 0 then
            flag_count = flag_count + 1
        end

        if r < 100 then
            chain_total = chain_total + 1

            if r % 13 == 0 then
                nested_count = nested_count + 1
            end
        elseif r < 400 then
            chain_total = chain_total + 2
        elseif r < 700 then
            chain_total = chain_total + 3
        else
            chain_total = chain_total + 4
        end

        if r % 97 == 0 then
            continue_count = continue_count + 1
            goto continue
        end

        do
            local k = r % 5

            if k == 0 then
                match_total = match_total + 10
            elseif k == 1 then
                match_total = match_total + 20
            elseif k == 2 then
                match_total = match_total + 30
            elseif k == 3 then
                match_total = match_total + 40
            else
                match_total = match_total + 50
            end

            local bonus
            if r % 2 == 0 then bonus = 1 else bonus = 2 end

            expr_sum = expr_sum + bonus
        end

        if i % 1000 == 0 then
            local ok, result = pcall(risky, r)

            if ok then
                fault_count = fault_count + result
            else
                fault_count = fault_count + 1
            end
        end

        ::continue::
    end

    return flag_count + chain_total * 3 + nested_count * 7 + continue_count * 11 +
        expr_sum * 13 + match_total * 17 + fault_count * 19
end

print(run())

File: bench/branches.js
// The JavaScript twin of branches.sl.
//
// JavaScript has no `match`, so the twin is a `switch`, which an engine is free to compile to a jump
// table or a chain of comparisons as it sees fit -- the same freedom `dispatch.js` takes. `continue`
// and `try`/`catch` are direct: JavaScript's own words for the same thing. The `if`-as-expression
// becomes JavaScript's `?:`, which is the nearest thing it has to an `if` that answers a value.
//
// Every JavaScript number is a double, but every value here stays an ordinary integer under 2^31 and
// the one product taken (`seed * 16807`) stays under 2^53, so this is exact integer arithmetic and
// prints the same answer as the other three.

function risky(r) {
    if (r === 0) throw new Error("boom");
    return 0;
}

function run() {
    let seed = 1;
    let i = 0;
    let flagCount = 0;
    let chainTotal = 0;
    let nestedCount = 0;
    let continueCount = 0;
    let exprSum = 0;
    let matchTotal = 0;
    let faultCount = 0;

    while (i < 2000000) {
        seed = (seed * 16807) % 2147483647;

        const r = seed % 1000;

        i = i + 1;

        if (r % 7 === 0) {
            flagCount = flagCount + 1;
        }

        if (r < 100) {
            chainTotal = chainTotal + 1;

            if (r % 13 === 0) {
                nestedCount = nestedCount + 1;
            }
        } else if (r < 400) {
            chainTotal = chainTotal + 2;
        } else if (r < 700) {
            chainTotal = chainTotal + 3;
        } else {
            chainTotal = chainTotal + 4;
        }

        if (r % 97 === 0) {
            continueCount = continueCount + 1;
            continue;
        }

        const k = r % 5;

        switch (k) {
            case 0: matchTotal = matchTotal + 10; break;
            case 1: matchTotal = matchTotal + 20; break;
            case 2: matchTotal = matchTotal + 30; break;
            case 3: matchTotal = matchTotal + 40; break;
            default: matchTotal = matchTotal + 50; break;
        }

        const bonus = r % 2 === 0 ? 1 : 2;

        exprSum = exprSum + bonus;

        if (i % 1000 === 0) {
            try {
                faultCount = faultCount + risky(r);
            } catch (e) {
                faultCount = faultCount + 1;
            }
        }
    }

    return flagCount + chainTotal * 3 + nestedCount * 7 + continueCount * 11 +
        exprSum * 13 + matchTotal * 17 + faultCount * 19;
}

console.log(run());
